File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math
from collections import defaultdict

def solution(fees, records):
    basic_time, basic_rate, unit_time, unit_rate = fees
    parking = {}
    total_time = defaultdict(int)

    for record in records:
        time_str, car, status = record.split()
        h, m = map(int, time_str.split(":"))
        minutes = h * 60 + m

        if status == "IN":
            parking[car] = minutes
        else:
            total_time[car] += minutes - parking.pop(car)

    for car, in_time in parking.items():
        total_time[car] += 1439 - in_time 

    result = []
    for car in sorted(total_time):
        time = total_time[car]
        if time <= basic_time:
            result.append(basic_rate)
        else:
            extra = math.ceil((time - basic_time) / unit_time)
            result.append(basic_rate + extra * unit_rate)

    return result

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
    const [basicTime, basicRate, unitTime, unitRate] = fees;
    
    let parking = new Map();
    let totalTime = new Map();
    
    for(const record of records){
        const [timeStr, car, type] = record.split(" ");
        const [h, m] = timeStr.split(":").map(Number);
        const time = 60*h + m;
        
        if(type === "IN"){
            parking.set(car, time);
        } else {
            const inTime = parking.get(car);
            const duration = time - inTime;
            totalTime.set(car, (totalTime.get(car) || 0) + duration);
            parking.delete(car);
        }
    }
    
    for(const [car, inTime] of parking.entries()){
        const duration = 1439 - inTime;
        totalTime.set(car, (totalTime.get(car) || 0) + duration);
    }
    
    const result = Array.from(totalTime.entries())
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([car, time]) => {
        if(time <= basicTime) return basicRate;
        return basicRate + Math.ceil((time - basicTime) / unitTime) * unitRate;
    });
    
    return result;
}